scalding-hadoop-test/src/test/scala/com/twitter/scalding/platform/PlatformTest.scala

```diff
@@ -154,6 +154,94 @@ class TypedPipeJoinWithDescriptionJob(args: Args) extends Job(args) {
+  val yMap = y.map( p => (p._1, p._2.toUpperCase)).forceToDisk
+
+  x.hashJoin(yMap)
+   .withDescription("hashJoin")
+   .write(TypedTsv[(Int, (Int, String))]("output"))
+}
+
+class TypedPipeHashJoinWithForceToDiskFilterJob(args: Args) extends Job(args) {
+  val x = TypedPipe.from[(Int, Int)](List((1, 1)))
+  val y = TypedPipe.from[(Int, String)](List((1, "first")))
+
+  //trivial transform and forceToDisk followed by filter on rhs
+  val yFilter = y.map( p => (p._1, p._2.toUpperCase)).forceToDisk.filter(p => p._1 == 1)
+
+  x.hashJoin(yFilter)
+   .withDescription("hashJoin")
+   .write(TypedTsv[(Int, (Int, String))]("output"))
+}
+
+class TypedPipeHashJoinWithForceToDiskWithComplete(args: Args) extends Job(args) {
+  val x = TypedPipe.from[(Int, Int)](List((1, 1)))
+  val y = TypedPipe.from[(Int, String)](List((1, "first")))
+
+  //trivial transform and forceToDisk followed by WithComplete on rhs
+  val yComplete = y.map( p => (p._1, p._2.toUpperCase)).forceToDisk.onComplete(() => println("step complete"))
+
+  x.hashJoin(yComplete)
+   .withDescription("hashJoin")
+   .write(TypedTsv[(Int, (Int, String))]("output"))
+}
+
+class TypedPipeHashJoinWithForceToDiskMapJob(args: Args) extends Job(args) {
+  val x = TypedPipe.from[(Int, Int)](List((1, 1)))
+  val y = TypedPipe.from[(Int, String)](List((1, "first")))
+
+  //trivial transform and forceToDisk followed by map on rhs
+  val yMap = y.map( p => (p._1, p._2.toUpperCase)).forceToDisk.map(p => (p._1, p._2.toLowerCase))
+
+  x.hashJoin(yMap)
+   .withDescription("hashJoin")
+   .write(TypedTsv[(Int, (Int, String))]("output"))
+}
+
+class TypedPipeHashJoinWithGroupByJob(args: Args) extends Job(args) {
+  val x = TypedPipe.from[(String, Int)](Tsv("input1", ('x1, 'y1)), Fields.ALL)
+  val y = Tsv("input2", ('x2, 'y2))
+
+  val yGroup = y.groupBy('x2){p => p}
+  val yTypedPipe = TypedPipe.from[(String, Int)](yGroup, Fields.ALL)
+
+  x.hashJoin(yTypedPipe)
+   .withDescription("hashJoin")
+   .write(TypedTsv[(String, (Int, Int))]("output"))
+}
+
+class TypedPipeHashJoinWithCoGroupJob(args: Args) extends Job(args) {
+  val x = TypedPipe.from[(Int, Int)](List((1, 1)))
+  val in0 = Tsv("input0").read.mapTo((0, 1) -> ('x0, 'a)) { input: (Int, Int) => input }
+  val in1 = Tsv("input1").read.mapTo((0, 1) -> ('x1, 'b)) { input: (Int, Int) => input }
+
+  val coGroupPipe = in0.coGroupBy('x0) {
+    _.coGroup('x1, in1, OuterJoinMode)
+  }
+
+  val coGroupTypedPipe = TypedPipe.from[(Int, Int, Int)](coGroupPipe, Fields.ALL)
+  val coGroupTuplePipe = coGroupTypedPipe.map{ case(a,b,c) => (a, (b,c)) }
+  x.hashJoin(coGroupTuplePipe)
+   .withDescription("hashJoin")
+   .write(TypedTsv[(Int, (Int, (Int, Int)))]("output"))
+}
+
+class TypedPipeHashJoinWithEveryJob(args: Args) extends Job(args) {
+  val x = TypedPipe.from[(Int, String)](Tsv("input1", ('x1, 'y1)), Fields.ALL)
+  val y = Tsv("input2", ('x2, 'y2)).groupBy('x2) {
+    _.foldLeft('y2 -> 'y2)(0){ (b: Int, a: Int) => b + a}
```

Comment:
Hmm seems like the line you've listed does not create an `Every` pipe. I end up with an Each -> source Pipe. Not super familiar with this but from what I can make out debugging, seems like when we do typedPipe.group we end up with a Grouped object instead of a GroupBuilder. The foldLeft on GroupBuilder is the only one that creates the Every (and not the one on Grouped). 

Seems like all the calls to create a new Every seem to be in GroupBuilder and that GroupBuilder is being used in RichPipe..